### PR TITLE
Enable CI via GitHub Actions

### PR DESCRIPTION
**Pull request by rupertnash**
See also the predecessor branch rupertnash/gh-actions for a slightly different way of running the jobs, but I like this way better.

TODO:

- [x] Fix failing tests (I guess they've crept in while CI is down?)
- [x] Remove Travis bits

Comments and extra commits welcome.

### CODE REVIEW

**Comment by rupertnash** [comment]
No idea what's up with all the failures.
E.g. this one: https://github.com/common-workflow-language/cwltool/runs/2041213091?check_suite_focus=true#step:9:47
```
Collecting bagit==1.6.4
36
  Downloading bagit-1.6.4.tar.gz (29 kB)
37
    ERROR: Command errored out with exit status 1:
38
     command: /home/runner/work/cwltool/cwltool/.tox/py36-unit/bin/python -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-nhxxplxk/bagit_2b92ad427998490e9c9f64445ce591c4/setup.py'"'"'; __file__='"'"'/tmp/pip-install-nhxxplxk/bagit_2b92ad427998490e9c9f64445ce591c4/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' egg_info --egg-base /tmp/pip-pip-egg-info-uoe1twjt
39
         cwd: /tmp/pip-install-nhxxplxk/bagit_2b92ad427998490e9c9f64445ce591c4/
40
    Complete output (7 lines):
41
    Traceback (most recent call last):
42
      File "<string>", line 1, in <module>
43
      File "/tmp/pip-install-nhxxplxk/bagit_2b92ad427998490e9c9f64445ce591c4/setup.py", line 19, in <module>
44
        long_description = readme.read()
45
      File "/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/encodings/ascii.py", line 26, in decode
46
        return codecs.ascii_decode(input, self.errors)[0]
47
    UnicodeDecodeError: 'ascii' codec can't decode byte 0xe2 in position 1111: ordinal not in range(128)
```

I've never seen that sort of one

**Comment by mr-c** [comment]
> No idea what's up with all the failures.
> E.g. this one: https://github.com/common-workflow-language/cwltool/runs/2041213091?check_suite_focus=true#step:9:47

> 
> I've never seen that sort of one

Maybe pin a newer version of pip and setuptools?
We can add that to the readme as being the minimum version as well.

What is the base image?

**Comment by rupertnash** [comment]
Base image? It's Ubuntu 20.04

When the workflow ran on my fork (exact same commit), the unit tests pass. See the equivalent here:
https://github.com/rupertnash/cwltool/runs/2041192514?check_suite_focus=true

**Comment by rupertnash** [comment]
I have no restrictions on actions on my forked repo, in case that might be the problem.

**Comment by rupertnash** [comment]
OK - I think it's something in the locale settings. Looking at the bagit@1.6.4 README.rst, line 44 does indeed contain a non-ASCII character (a "-" lookalike)
https://github.com/LibraryOfCongress/bagit-python/blob/v1.6.4/README.rst#L44

Bagit have fixed the bug in their setup.py, in 1.7.0 (https://github.com/LibraryOfCongress/concordia/issues/69)

Can we bump the bagit version? In anticipation of a "yes", I've pushed that change, but can easily reset.

File: release-test.sh [hunk 1]
```diff
@@ -5,6 +5,10 @@ set -x
 
 export LC_ALL=C
 
+echo "Begin environment"
+env
+echo "End environment"
+
 package=cwltool
 module=cwltool
 slug=${TRAVIS_PULL_REQUEST_SLUG:=common-workflow-language/cwltool}
```
Comment (mr-c) [inline release-test.sh:14]:
```suggestion
slug=${GITHUB_REPOSITORY:=common-workflow-language/cwltool}
```

File: .github/workflows/ci-tests.yml [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+    name: CWL spec conformance tests
+
+    runs-on: ubuntu-20.04
+
+    env:
+      py-semver: 3.8
+
+    strategy:
+      matrix:
+        cwl-version: [v1.0, v1.1, v1.2.0-dev4]
```
Comment (mr-c) [inline .github/workflows/ci-tests.yml:123]:
```suggestion
        cwl-version: [v1.0, v1.1, v1.2]
```

**Review by mr-c — CHANGES_REQUESTED** [review]
Looking great! Time to delete the Travis bits and rename files as appropriate?

File: .github/workflows/ci-tests.yml [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+          - py-ver-major: 3
+            py-ver-minor: 9
+            step: mypy
+
+    env:
+      py-semver: ${{ format('{0}.{1}', matrix.py-ver-major, matrix.py-ver-minor) }}
+      TOXENV: ${{ format('py{0}{1}-{2}', matrix.py-ver-major, matrix.py-ver-minor, matrix.step) }}
+
+    steps:
+      - uses: actions/checkout@v2
```
Comment (mr-c) [inline .github/workflows/ci-tests.yml:32]:
```suggestion
      - uses: actions/checkout@v2
        with:
          fetch-depth: 0
```

As per https://github.com/python/importlib_metadata/pull/252/files

Comment (rupertnash) [reply .github/workflows/ci-tests.yml:32]: Good spot. I was very confused as to why this failed. Never forget to RTFM.

Comment (rupertnash) [reply .github/workflows/ci-tests.yml:32]: Actually, I think you mean to do this on line 92?

Comment (mr-c) [reply .github/workflows/ci-tests.yml:32]: > Good spot. I was very confused as to why this failed. Never forget to RTFM.

I just did a web search for the error message :-P

Comment (rupertnash) [reply .github/workflows/ci-tests.yml:32]: It's not resolved, but by replaying the git commands from CI I can reproduce the error locally. The diff command is incorrect cos `main` isn't a local branch. Is there any problem setting `--compare-branch=origin/main` in the Makefile for `diff_pydocstyle_report`?

Comment (rupertnash) [reply .github/workflows/ci-tests.yml:32]: Or I guess I could just add `git branch main origin/main` before running?

File: .github/workflows/ci-tests.yml [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+      - name: Test with tox
+        run: tox
+
+  conformance_tests:
+    name: CWL spec conformance tests
+
+    runs-on: ubuntu-20.04
+
+    env:
+      py-semver: 3.8
```
Comment (mr-c) [inline .github/workflows/ci-tests.yml:125]: Here as well

File: .github/workflows/ci-tests.yml [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+          version: ${{ matrix.cwl-version }}
+        run: ./conformance-test.sh
+
+  release_test:
+    name: CWL release test
+
+    runs-on: ubuntu-20.04
+
+    env:
+      py-semver: 3.7
```
Comment (mr-c) [inline .github/workflows/ci-tests.yml:156]: Once more

**Comment by rupertnash** [comment]
Quit asking me to make you CI better, rather than just migrating! 🙊

**Review by mr-c — APPROVED** [review]
Many, many thanks @rupertnash !

**Pull request merged**